feat(ci): Add new command tag_branch

kskewes-sf opened the pull request:
Builds on #189 

- tag `master` branch with next semVer minor version.
- tag all other branches (eg: `release-*`) with next semVer patch version.

Because of auto-bump GitHub Actions kicking off a fresh round of tag
bumps across some service repositories we can't just use the freshly
created tag here in BOM generation.

kskewes-sf committed:
feat(ci): Add new command tag_branch

- tag `master` branch with next semVer minor version.
- tag all other branches (eg: `release-*`) with next semVer patch version.
- process repositories for core Halyard supported services plus keel & spin CLI
- add usage instructions to README as we are manually executing this for
  now until automation is done.
- downgrade build_bom untagged HEAD warning log to debug because it is
  now just noise in the new release cut flow (`tag_branch` takes care
  of tagging `HEAD`).

Because of auto-bump GitHub Actions kicking off a fresh round of tag
bumps across some service repositories we can't just use the freshly
created tag here in BOM generation.

kskewes-sf commented on dev/buildtool/__init__.py:
```diff
@@ -15,6 +15,7 @@
     "orca",
     "rosco",
 ]
+SPINNAKER_RUNNABLE_NON_CORE_REPOSITORY_NAMES = ["keel"]
```
Halyard doesn't explicitly support `keel` so it hasn't been added to the list above.

However we still want to create tags and release-* branches.
We can't easily add it to the changelog though without adding it to the BOM which requires some Halyard validation.

kskewes-sf commented on dev/buildtool/git_support.py:
```diff
@@ -863,7 +863,7 @@ def find_newest_tag_and_common_commit_from_id(self, git_dir, commit_id):
         if tag_commit == commit_id:
             logging.debug("Commit %s is already tagged with %s", tag_commit, tag)
         else:
-            logging.warning(
```
This was previously an important warning when building a BOM without `tag_branch` command help.
However `tag_branch` calls this method and based on the result will generate a new tag or not, taking care of the situation for us.

kskewes-sf commented on dev/buildtool/source_commands.py:
```diff
@@ -138,6 +141,97 @@ def init_argparser(self, parser, defaults):
+        all_names.extend(SPIN_REPOSITORY_NAMES)
+        super().__init__(factory, options, source_repository_names=all_names)
+
+    def _do_repository(self, repository):
```
For each repository in `all_names` above, perform the following:
1. clone repo
2. get latest tag & its commit ID
3. compare to HEAD commit ID, if it matches do nothing for this repo.
4. generate a new minor or patch tag branch depending
5. push tag to git remote

kskewes-sf committed:
feat(ci): simplify logging in tag_branch, remove broken flags

kskewes-sf commented:
Example run to add a new patch tag to `release-1.27.x` branch on my forks.
```
$ ./dev/buildtool.sh tag_branch --git_branch release-1.27.x --github_owner kskewes-sf --git_never_push true W 10:56:15.335 [MainThread.2378142] Monitoring is disabled
I 10:56:15.339 [MainThread.2378142] Mapping 12/['clouddriver', 'deck', 'echo', 'fiat', 'front50', 'gate', 'igor', 'kayenta', 'orca', 'rosco', 'keel', 'spin']
I 10:56:15.340 [Thread-1.2378142] tag_branch processing clouddriver
I 10:56:15.340 [Thread-2.2378142] tag_branch processing deck
I 10:56:15.340 [Thread-3.2378142] tag_branch processing echo
I 10:56:15.340 [Thread-4.2378142] tag_branch processing fiat
I 10:56:15.342 [Thread-5.2378142] tag_branch processing front50
I 10:56:15.343 [Thread-6.2378142] tag_branch processing gate
I 10:56:15.347 [Thread-7.2378142] tag_branch processing igor
I 10:56:15.348 [Thread-8.2378142] tag_branch processing kayenta
I 10:56:15.350 [Thread-9.2378142] tag_branch processing orca
I 10:56:15.350 [Thread-10.2378142] tag_branch processing rosco
I 10:56:15.351 [Thread-11.2378142] tag_branch processing keel
I 10:56:15.352 [Thread-12.2378142] tag_branch processing spin
I 10:56:18.779 [Thread-12.2378142] Cloned https://github.com/kskewes-sf/spin into source_code/tag_branch
I 10:56:18.802 [Thread-12.2378142] spin latest tag: v1.27.1 not at HEAD, generating next tag: v1.27.2
W 10:56:18.806 [Thread-12.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/spin" push origin v1.27.2
I 10:56:18.807 [Thread-12.2378142] tag_branch finished spin
I 10:56:19.232 [Thread-10.2378142] Cloned https://github.com/kskewes-sf/rosco into source_code/tag_branch
I 10:56:19.257 [Thread-10.2378142] rosco latest tag: v1.7.4 not at HEAD, generating next tag: v1.7.5
W 10:56:19.261 [Thread-10.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/rosco" push origin v1.7.5
I 10:56:19.261 [Thread-10.2378142] tag_branch finished rosco
I 10:56:19.740 [Thread-3.2378142] Cloned https://github.com/kskewes-sf/echo into source_code/tag_branch
I 10:56:19.769 [Thread-3.2378142] echo latest tag: v2.32.3 not at HEAD, generating next tag: v2.32.4
W 10:56:19.774 [Thread-3.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/echo" push origin v2.32.4
I 10:56:19.774 [Thread-3.2378142] tag_branch finished echo
I 10:56:20.450 [Thread-4.2378142] Cloned https://github.com/kskewes-sf/fiat into source_code/tag_branch
I 10:56:20.475 [Thread-4.2378142] fiat latest tag: v1.28.4 not at HEAD, generating next tag: v1.28.5
W 10:56:20.479 [Thread-4.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/fiat" push origin v1.28.5
I 10:56:20.480 [Thread-4.2378142] tag_branch finished fiat
I 10:56:20.483 [Thread-7.2378142] Cloned https://github.com/kskewes-sf/igor into source_code/tag_branch
I 10:56:20.508 [Thread-7.2378142] igor latest tag: v4.6.4 not at HEAD, generating next tag: v4.6.5
W 10:56:20.513 [Thread-7.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/igor" push origin v4.6.5
I 10:56:20.513 [Thread-7.2378142] tag_branch finished igor
I 10:56:21.580 [Thread-8.2378142] Cloned https://github.com/kskewes-sf/kayenta into source_code/tag_branch
I 10:56:21.601 [Thread-8.2378142] kayenta latest tag: v2.31.2 not at HEAD, generating next tag: v2.31.3
W 10:56:21.604 [Thread-8.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/kayenta" push origin v2.31.3
I 10:56:21.605 [Thread-8.2378142] tag_branch finished kayenta
I 10:56:23.563 [Thread-6.2378142] Cloned https://github.com/kskewes-sf/gate into source_code/tag_branch
I 10:56:23.590 [Thread-6.2378142] gate latest tag: v6.54.2 not at HEAD, generating next tag: v6.54.3
W 10:56:23.594 [Thread-6.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/gate" push origin v6.54.3
I 10:56:23.594 [Thread-6.2378142] tag_branch finished gate
I 10:56:24.344 [Thread-5.2378142] Cloned https://github.com/kskewes-sf/front50 into source_code/tag_branch
I 10:56:24.370 [Thread-5.2378142] front50 latest tag: v2.23.3 not at HEAD, generating next tag: v2.23.4
W 10:56:24.374 [Thread-5.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/front50" push origin v2.23.4
I 10:56:24.374 [Thread-5.2378142] tag_branch finished front50
I 10:56:34.768 [Thread-11.2378142] Cloned https://github.com/kskewes-sf/keel into source_code/tag_branch
I 10:56:34.794 [Thread-11.2378142] keel latest tag: v1.1.2 not at HEAD, generating next tag: v1.1.3
W 10:56:34.799 [Thread-11.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/keel" push origin v1.1.3
I 10:56:34.800 [Thread-11.2378142] tag_branch finished keel
I 10:56:41.154 [Thread-9.2378142] Cloned https://github.com/kskewes-sf/orca into source_code/tag_branch
I 10:56:41.190 [Thread-9.2378142] orca latest tag: v8.18.4 not at HEAD, generating next tag: v8.18.5
W 10:56:41.195 [Thread-9.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/orca" push origin v8.18.5
I 10:56:41.195 [Thread-9.2378142] tag_branch finished orca
I 10:56:44.396 [Thread-1.2378142] Cloned https://github.com/kskewes-sf/clouddriver into source_code/tag_branch
I 10:56:44.435 [Thread-1.2378142] clouddriver latest tag: v5.74.3 not at HEAD, generating next tag: v5.74.4
W 10:56:44.439 [Thread-1.2378142] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/clouddriver" push origin v5.74.4
I 10:56:44.439 [Thread-1.2378142] tag_branch finished clouddriver
I 10:56:48.666 [Thread-2.2378142] Cloned https://github.com/kskewes-sf/deck into source_code/tag_branch
I 10:56:48.687 [Thread-2.2378142] deck HEAD commit: c1902676fe7cc0dbcc6327ad42cd84492f242b27 already tagged at: v3.8.2. Skipping.
I 10:56:48.687 [Thread-2.2378142] tag_branch finished deck
I 10:56:48.687 [MainThread.2378142] Finished mapping
W 10:56:48.689 [MainThread.2378142] Monitoring disabled -- dont push incremental metrics.
W 10:56:48.689 [MainThread.2378142] Monitoring disabled -- dont push final metrics.
```

kskewes-sf commented:
And for `master` branch:
```
$ ./dev/buildtool.sh tag_branch --git_branch master --github_owner kskewes-sf --git_never_push true 
W 10:57:16.081 [MainThread.2378534] Monitoring is disabled
I 10:57:16.086 [MainThread.2378534] Mapping 12/['clouddriver', 'deck', 'echo', 'fiat', 'front50', 'gate', 'igor', 'kayenta', 'orca', 'rosco', 'keel', 'spin']
I 10:57:16.086 [Thread-1.2378534] tag_branch processing clouddriver
I 10:57:16.086 [Thread-2.2378534] tag_branch processing deck
I 10:57:16.087 [Thread-3.2378534] tag_branch processing echo
I 10:57:16.090 [Thread-4.2378534] tag_branch processing fiat
I 10:57:16.090 [Thread-5.2378534] tag_branch processing front50
I 10:57:16.091 [Thread-6.2378534] tag_branch processing gate
I 10:57:16.092 [Thread-7.2378534] tag_branch processing igor
I 10:57:16.097 [Thread-8.2378534] tag_branch processing kayenta
I 10:57:16.098 [Thread-9.2378534] tag_branch processing orca
I 10:57:16.100 [Thread-10.2378534] tag_branch processing rosco
I 10:57:16.105 [Thread-11.2378534] tag_branch processing keel
I 10:57:16.106 [Thread-12.2378534] tag_branch processing spin
I 10:57:19.803 [Thread-4.2378534] Cloned https://github.com/kskewes-sf/fiat into source_code/tag_branch
I 10:57:19.829 [Thread-4.2378534] fiat latest tag: v1.31.0 not at HEAD, generating next tag: v1.32.0
W 10:57:19.834 [Thread-4.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/fiat" push origin v1.32.0
I 10:57:19.835 [Thread-4.2378534] tag_branch finished fiat
I 10:57:19.837 [Thread-10.2378534] Cloned https://github.com/kskewes-sf/rosco into source_code/tag_branch
I 10:57:19.843 [Thread-12.2378534] Cloned https://github.com/kskewes-sf/spin into source_code/tag_branch
I 10:57:19.862 [Thread-10.2378534] rosco latest tag: v1.10.0 not at HEAD, generating next tag: v1.11.0
W 10:57:19.867 [Thread-10.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/rosco" push origin v1.11.0
I 10:57:19.867 [Thread-10.2378534] tag_branch finished rosco
I 10:57:19.871 [Thread-12.2378534] spin latest tag: v1.27.0 not at HEAD, generating next tag: v1.28.0
W 10:57:19.876 [Thread-12.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/spin" push origin v1.28.0
I 10:57:19.876 [Thread-12.2378534] tag_branch finished spin
I 10:57:22.012 [Thread-5.2378534] Cloned https://github.com/kskewes-sf/front50 into source_code/tag_branch
I 10:57:22.035 [Thread-5.2378534] front50 latest tag: v2.25.0 not at HEAD, generating next tag: v2.26.0
W 10:57:22.039 [Thread-5.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/front50" push origin v2.26.0
I 10:57:22.039 [Thread-5.2378534] tag_branch finished front50
I 10:57:23.604 [Thread-3.2378534] Cloned https://github.com/kskewes-sf/echo into source_code/tag_branch
I 10:57:23.630 [Thread-3.2378534] echo latest tag: v2.34.0 not at HEAD, generating next tag: v2.35.0
W 10:57:23.633 [Thread-3.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/echo" push origin v2.35.0
I 10:57:23.634 [Thread-3.2378534] tag_branch finished echo
I 10:57:25.680 [Thread-7.2378534] Cloned https://github.com/kskewes-sf/igor into source_code/tag_branch
I 10:57:25.704 [Thread-7.2378534] igor latest tag: v4.7.0 not at HEAD, generating next tag: v4.8.0
W 10:57:25.708 [Thread-7.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/igor" push origin v4.8.0
I 10:57:25.708 [Thread-7.2378534] tag_branch finished igor
I 10:57:27.426 [Thread-6.2378534] Cloned https://github.com/kskewes-sf/gate into source_code/tag_branch
I 10:57:27.454 [Thread-6.2378534] gate latest tag: v6.55.0 not at HEAD, generating next tag: v6.56.0
W 10:57:27.458 [Thread-6.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/gate" push origin v6.56.0
I 10:57:27.458 [Thread-6.2378534] tag_branch finished gate
I 10:57:32.982 [Thread-11.2378534] Cloned https://github.com/kskewes-sf/keel into source_code/tag_branch
I 10:57:33.007 [Thread-11.2378534] keel latest tag: v1.2.0 not at HEAD, generating next tag: v1.3.0
W 10:57:33.011 [Thread-11.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/keel" push origin v1.3.0
I 10:57:33.011 [Thread-11.2378534] tag_branch finished keel
I 10:57:34.448 [Thread-8.2378534] Cloned https://github.com/kskewes-sf/kayenta into source_code/tag_branch
I 10:57:34.470 [Thread-8.2378534] kayenta latest tag: v2.32.0 not at HEAD, generating next tag: v2.33.0
W 10:57:34.473 [Thread-8.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/kayenta" push origin v2.33.0
I 10:57:34.474 [Thread-8.2378534] tag_branch finished kayenta
I 10:57:39.512 [Thread-9.2378534] Cloned https://github.com/kskewes-sf/orca into source_code/tag_branch
I 10:57:39.549 [Thread-9.2378534] orca latest tag: v8.24.0 not at HEAD, generating next tag: v8.25.0
W 10:57:39.553 [Thread-9.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/orca" push origin v8.25.0
I 10:57:39.553 [Thread-9.2378534] tag_branch finished orca
I 10:57:53.267 [Thread-2.2378534] Cloned https://github.com/kskewes-sf/deck into source_code/tag_branch
I 10:57:53.331 [Thread-2.2378534] deck latest tag: v3.9.0 not at HEAD, generating next tag: v3.10.0
W 10:57:53.335 [Thread-2.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/deck" push origin v3.10.0
I 10:57:53.335 [Thread-2.2378534] tag_branch finished deck
I 10:57:57.388 [Thread-1.2378534] Cloned https://github.com/kskewes-sf/clouddriver into source_code/tag_branch
I 10:57:57.425 [Thread-1.2378534] clouddriver latest tag: v5.76.0 not at HEAD, generating next tag: v5.77.0
W 10:57:57.429 [Thread-1.2378534] SKIP pushing tag because --git_never_push=true.
Command would have been: git -C "source_code/tag_branch/clouddriver" push origin v5.77.0
I 10:57:57.430 [Thread-1.2378534] tag_branch finished clouddriver
I 10:57:57.430 [MainThread.2378534] Finished mapping
W 10:57:57.432 [MainThread.2378534] Monitoring disabled -- dont push incremental metrics.
W 10:57:57.432 [MainThread.2378534] Monitoring disabled -- dont push final metrics.
```

kskewes-sf commented on dev/buildtool/source_commands.py:
```diff
@@ -138,6 +141,97 @@ def init_argparser(self, parser, defaults):
+
+        all_names = list(SPINNAKER_RUNNABLE_REPOSITORY_NAMES)
+        all_names.extend(SPINNAKER_RUNNABLE_NON_CORE_REPOSITORY_NAMES)
+        all_names.extend(SPIN_REPOSITORY_NAMES)
```
Currently for `spin` CLI the `master` branch tag (`v1.27.0`) corresponds with the Spinnaker Release version (1.27.0).

If we always include `spin` CLI here then we might end up bumping the `master` branch out of sync with the Spinnaker Release version. 
However we won't have to tag the repo manually each Spinnaker Release.

We could manually tag `spin` `master` branch and have `buildtool` only tag `spin` non-`master` branches?

dbyron-sf replied:
Not sure I understand the struggle here.  Is it that we haven't been cutting release-1.xy.x branches for spin?

kskewes-sf replied:
We do have `release-1.xy.z` branches for `spin` but the tag on `master` branch currently matches the Spinnaker Release Version.
For other services the tag on `master` branch is not connected (by chance or otherwise) to a Spinnaker Release version, eg: gate might be `v6.5.0`

dbyron-sf replied:
Is it overkill to add some other list of repos that follow the spinnaker version?  And then treat that list of repos as special in the code?  It does seem a shame to sacrifice the version matching...

kskewes-sf replied:
It's simple enough to add another constant but there's a catch.

`tag_branch` needs a `spinnaker_version` flag in order to tag `spin`. 
I'm not sure we want to add this so let's kick `spin` down the road a little until we tackle a revision of the `create_release_branch` command (which has `spinnaker_version` flag).

kskewes-sf replied:
Removed `spin` from repository targets for `tag_branch` in https://github.com/spinnaker/buildtool/pull/190/commits/60d5f4f07e99451c2af95c909f1e10966fd8ed8e#diff-09003a65273be87edc536caabb89a3977ee30512f1c7931745ab1bf0695579e9L154

dbyron-sf commented on README.md:
```diff
@@ -54,6 +54,44 @@ FAILED (errors=4)
+At time of writing, tagging designated (`master` and `release-*`) branches will:
+
+1. (on `master`) provide a commit SHA to branch at for a new Spinnaker Release,
+   e.g: `1.28.0`.
```
I'm not sure 1.28.0 is a great example here....since typically microservice versions don't line up with overall spinnaker versions.

kskewes-sf replied:
Thanks, have reworded, how does this sound?  https://github.com/spinnaker/buildtool/pull/190/commits/2ced9a75c5a496beb357387de79b1596ecb7cc5c#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R69

> 1. (on `master`) provide a commit SHA to create new `release-*` branches at.

dbyron-sf replied:
Yeah that works.  The grammar weenie in me bristles a little with `at` ending a sentence, but
```
(on master) provide a commit SHA with which to create a new `release-*` branch
```
is a bit of a mouthful.

I do see another potential place to improve...the text before this says
```
At time of writing, tagging designated (`master` and `release-*`) branches will:
```
So that's talking about more than one branch (`branches`)....but here we're talking about one commit SHA and so I figure that's also one branch....

But maybe it's better if it all talks about one branch?

dbyron-sf commented on README.md:
````diff
@@ -54,6 +54,44 @@ FAILED (errors=4)
+   and build a new set of artifacts.
+
+```
+./dev/buildtool.sh tag_branches \
````
nit: the PR description and commit message say `tag_branch`, but we've got `tag_branches` here.

dbyron-sf replied:
Looks like the command is actually named `tag_branch` in the code, so:
```suggestion
./dev/buildtool.sh tag_branch \
```

dbyron-sf commented on README.md:
````diff
@@ -54,6 +54,44 @@ FAILED (errors=4)
+  --only_repositories clouddriver \
+  --git_never_push true
+```
+
````
Is the TODO below still relevant?
```
TODO: Check that this walks back to last tag and doesn't create branches from
HEAD.
```
maybe a previous PR already addressed this?  Perhaps https://github.com/spinnaker/buildtool/pull/185?

kskewes-sf replied:
Yes still relevant. The [code](https://github.com/spinnaker/buildtool/blob/master/dev/buildtool/spinnaker_commands.py#L141) currently doesn't checkout a tag before creating a branch.

Follow up PR will sort out.

dbyron-sf commented on dev/buildtool/source_commands.py:
```diff
@@ -138,6 +141,81 @@ def init_argparser(self, parser, defaults):
+    def _do_repository(self, repository):
+        """Implements RepositoryCommandProcessor interface."""
+        head_commit_id = self.__git.query_local_repository_commit_id(repository.git_dir)
+        logging.debug("%s HEAD commit: %s", repository.name, head_commit_id)
```
include the branch name in the log message, here and below?

kskewes-sf replied:
Good idea, especially for when a GHA might run this.

Done in: https://github.com/spinnaker/buildtool/pull/190/commits/2ced9a75c5a496beb357387de79b1596ecb7cc5c
Example below: 
```
I 10:36:13.505 [Thread-4.2455274] fiat release-1.27.x branch HEAD commit: 52a8f5204dc12b693b8eac5ff6126759c119684a already tagged at: v1.28.5. Skipping.
```

dbyron-sf commented on README.md:
```diff
@@ -54,6 +54,44 @@ FAILED (errors=4)
+git_branch=master
+fork_owner=<you>
+
+./dev/buildtool.sh tag_branches \
```
```suggestion
./dev/buildtool.sh tag_branch \
```

kskewes-sf committed:
chore(ci): Correct README instruction typos for tag_branch

Co-authored-by: David Byron <82477955+dbyron-sf@users.noreply.github.com>

kskewes-sf committed:
chore(ci): Log branch name during tag_branch actions

kskewes-sf committed:
chore(ci): Reword tag_branch README instructions

kskewes-sf committed:
chore(ci): Skip tagging spin repo for now

We want to maintain `spin` repo's `master` branch tagging alignment with
the Spinnaker Release versions.

This is probably going to require tagging during release branch creation
as the `tag_branch` command doesn't (and shouldn't?) require
a `spinnaker_version` flag.
